src/outputs/manifest_output.py

```diff
@@ -0,0 +1,186 @@
+import json
+import os
+from datetime import datetime
+
+
+class ManifestError(Exception):
+    pass
+
+
+class Manifest:
+    """
+    An outgoing file transfer manifest. Accumulates file metadata before
+        writing a JSON manifest file.
+
+    Used by Apache NiFi for data integrity checks before initiating data
+        transfers.
+
+    Attributes
+    ==========
+    outgoing_directory
+        location to write files to
+    pipeline_run_datetime
+        datetime of the current pipeline run, used to version outputs
+    dry_run
+        when True, cleans up output files after a successful run
+    """
+
+    def __init__(
+        self,
+        outgoing_directory: str,
+        pipeline_run_datetime: datetime,
+        delete_file_func: callable,
+        md5sum_func: callable,
+        stat_size_func: callable,
+        isdir_func: callable,
+        isfile_func: callable,
+        read_header_func: callable,
+        string_to_file_func: callable,
+        dry_run: bool = False,
+    ):
+        self.outgoing_directory = outgoing_directory
+        if not isdir_func(outgoing_directory):
+            raise ManifestError(
+                f"Outgoing directory does not exist: {self.outgoing_directory}"
+            )
+
+        if not isinstance(pipeline_run_datetime, datetime):
+            raise ManifestError("Pipeline run datetime must be a datetime object.")
+
+        self.manifest_datetime = pipeline_run_datetime.strftime("%Y%m%d_%H%M%S")
+
+        self.manifest_file_path = os.path.join(
+            outgoing_directory, (self.manifest_datetime + "_manifest.json")
+        )
+        self.manifest: dict = {"files": []}
+        self.written = False
+        self.invalid_headers: list = []
+        self.dry_run = dry_run
+
+        # Functions
+        self.delete_file = delete_file_func
+        self.md5sum = md5sum_func
+        self.stat_size = stat_size_func
+        self.isdir = isdir_func
+        self.isfile = isfile_func
+        self.read_header = read_header_func
+        self.string_to_file = string_to_file_func
+
+    def add_file(
```

Comment:
Suggest breaking up into further helper functions:
```suggestion
    def add_file(
        self,
        relative_file_path: str,
        column_header: str,
        validate_col_name_length: bool = True,
        sep: str = ",",
):
    if ".." in relative_file_path:
        raise ManifestError(
            (
                F"File must be in the subdirectory of the outgoing directory: {relative_file_path}"
            )
        )

    absolute_file_path = os.path.join(self.outgoing_directory, relative_file_path)

    if not self.isfile(absolute_file_path):
        raise ManifestError(
            (
                f"Cannot add file to manifest, file does not exist {absolute_file_path}"
            )
        )

    file_header_string = self.read_header(absolute_file_path)
    file_header_list = self._cleanup_and_split_header(file_header_string, sep)

    if not self._validate_column_headers(column_header, file_header_string, file_header_list):
        return  # Validation failed, do not add the file to the manifest

    if validate_col_name_length:
        self._check_column_name_length(absolute_file_path, file_header_list)

    file_manifest = {
        "file": os.path.basename(relative_file_path),
        "subfolder": os.path.dirname(relative_file_path),
        "sizeBytes": self.stat_size(absolute_file_path),
        "md5sum": self.md5sum(absolute_file_path),
        "header": column_header,
    }
    self.manifest["files"].append(file_manifest)

def _cleanup_and_split_header(self, file_header_string: str, sep: str) -> list:
    # Cleanup file_header_string because \n is appearing in it
    file_header_string = file_header_string.replace("\n", "")
    return file_header_string.split(sep)

def _validate_column_headers(
        self, expected_header: str, file_header_string: str, file_header_list: list
) -> bool:
    if file_header_string != expected_header:
        # Column headers in file do not match expected column headers
        column_header_list = expected_header.split(',')
        missing_columns = set(column_header_list) - set(file_header_list)
        additional_columns = set(file_header_list) - set(column_header_list)

        self.invalid_headers.append(
            f"File: {absolute_file_path}\n"
            f"Expected: {expected_header}\n"
            f"Got: {file_header_string}\n"
            f"Missing: {missing_columns}\n"
            f"Additional: {additional_columns}\n"
        )
        return False  # Validation failed
    return True  # Validation succeeded

def _check_column_name_length(self, absolute_file_path: str, file_header_list: list):
    col_above_max_len = [head for head in file_header_list if len(head) > 32]

    if col_above_max_len:
        self.invalid_headers.append(
            (
                f"File:{absolute_file_path}\n"
                f"These column names are exceeding the max char length "
                f"of 32: {col_above_max_len}\n"
            )
        )
```